File: BHAVRP_Python/cujae/inf/citi/om/assignment/others/distance/nearest_by_customer.py
import numpy as np
from typing import List
from ...classical.by_not_urgency import ByNotUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.solution.solution import Solution
from ....problem.solution.cluster import Cluster

class NearestByCustomer(ByNotUrgency):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        
    def to_clustering(self):
        self.initialize()
        self.assign()
        return self.finish()
    
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.total_customers = len(self.list_customers_to_assign)
        
    def assign(self):
        cost_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        current_customer_index = -1
        
        while self.list_customers_to_assign and self.list_clusters:
            current_customer_index += 1
            current_customer = self.list_customers_to_assign[0]
            request_customer = current_customer.get_request_customer()
            
            cost_col = cost_matrix[:, current_customer_index] 
            row = np.argmin(cost_col)
            col = current_customer_index
            
            id_depot = Problem.get_problem().get_list_id_depots()[row]
            depot = Problem.get_problem().get_depot_by_id_depot(id_depot)
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(depot)

            pos_cluster = self.find_cluster(id_depot, self.list_clusters)

            if pos_cluster != -1:
                cluster = self.list_clusters[pos_cluster]
                request_cluster = cluster.get_request_cluster()

                if capacity_depot >= (request_cluster + request_customer):
                    cluster.set_request_cluster(request_cluster + request_customer)
                    cluster.get_items_of_cluster().append(current_customer.get_id_customer())

                    capacity_depot -= request_customer
                    self.list_customers_to_assign.pop(0)
                    
                    if self.is_full_depot(self.list_customers_to_assign, cluster.get_request_cluster(), capacity_depot):
                        if cluster.get_items_of_cluster():
                            self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                    
                    cost_matrix[row, col] = float("inf")    
            
            if np.all(cost_matrix == float("inf")):
                self.solution.get_total_unassigned_items().append(current_customer.get_id_customer())
                self.list_customers_to_assign.pop(0)

    def finish(self):
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                self.solution.get_unassigned_items().append(customer.get_id_customer())

        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
        
        OSRMService.clear_distance_cache()
        
        return self.solution

File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/assignment/others/distance/NearestByCustomer.java
package cujae.inf.ic.om.assignment.others.distance;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.Cluster;
import cujae.inf.ic.om.problem.output.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

/*Clase que modela como asignar los clientes a los depósitos considerando el orden de la lista de clientes y se parte del criterio de cercanía a los depósitos*/
public class NearestByCustomer extends ByNotUrgency {

	private Solution solution = new Solution();	
	
	private ArrayList<Cluster> listClusters; 
	private ArrayList<Customer> listCustomersToAssign;
	private NumericMatrix costMatrix;
	
	public NearestByCustomer() {
		super();
	}

	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}
	
	@Override
	public void initialize() {
		listClusters = initializeClusters(); 
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		costMatrix = initializeCostMatrix(listCustomersToAssign, Problem.getProblem().getDepots(), distanceType);
	}

	@Override
	public void assign() {
		int idDepot = -1;
		int posDepot = -1;
		double capacityDepot = 0.0;

		int posCustomer = 0;
		double requestCustomer = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;
		
		RowCol rcBestDepot = null;
		int countTry = 0;  
		
		int totalCustomers = listCustomersToAssign.size();
		int totalDepots = Problem.getProblem().getDepots().size();
		
		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty()) && (!costMatrix.fullMatrix(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1), Double.POSITIVE_INFINITY)))
		{	
			rcBestDepot = costMatrix.indexLowerValue(totalCustomers, posCustomer, (totalCustomers + totalDepots - 1), posCustomer);
			
			posDepot = (rcBestDepot.getRow() - totalCustomers);
			idDepot = Problem.getProblem().getListIDDepots().get(posDepot);
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idDepot));
			
			requestCustomer = listCustomersToAssign.get(0).getRequestCustomer();
			
			costMatrix.setItem(rcBestDepot.getRow(), rcBestDepot.getCol(), Double.POSITIVE_INFINITY);
			
			posCluster = findCluster(idDepot, listClusters);
		
			if(posCluster != -1)
			{
				requestCluster = listClusters.get(posCluster).getRequestCluster();
				
				if(capacityDepot >= (requestCluster + requestCustomer)) 
				{
					requestCluster += requestCustomer;

					listClusters.get(posCluster).setRequestCluster(requestCluster);
					listClusters.get(posCluster).getItemsOfCluster().add(listCustomersToAssign.get(0).getIDCustomer());

					listCustomersToAssign.remove(0);
					posCustomer++;

					if(countTry != 0)
						countTry = 0;
				} 
				else 
				{
					countTry++;

					if(countTry >= listClusters.size())
					{
						countTry = 0;
						
						solution.getUnassignedItems().add(listCustomersToAssign.get(0).getIDCustomer());
						listCustomersToAssign.remove(0);
						posCustomer++;
					}
				}
				
				if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
				{
					if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
						solution.getClusters().add(listClusters.remove(posCluster));
					else
						listClusters.remove(posCluster);

					costMatrix.fillValue(rcBestDepot.getRow(), 0, rcBestDepot.getRow(), (totalCustomers - 1), Double.POSITIVE_INFINITY);
				}
			}
		}
	}
	
	@Override
	public Solution finish() {				
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));

		OSRMService.clearDistanceCache();
		
		return solution;
	}
}